server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecider.java

```diff
@@ -104,4 +106,69 @@ public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode n
             return decision;
         }
     }
+
+    /**
+     * Returns a {@link Decision} whether the given shard can be forced to the
+     * given node in the event that the shard's source node is being vacated.
+     * This allows nodes using a vacate-type node shutdown (replace/vacate) to
+     * override certain deciders in the interest of moving the shard away from
+     * a node that *must* be removed.
+     *
+     * It defaults to returning "NO" and must be overridden by deciders that
+     * opt-in to having their other NO decisions overridden while vacating.
+     */
+    public Decision canForceDuringVacate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return Decision.NO;
```

Comment:
I think we should reverse this, such that deciders have to opt-out, just like for `canForceAllocatePrimary`. That means that we do not inadvertently introduce a new decider that breaks REPLACE/VACATE.

It should respect THROTTLE though.

Reply:
is your thought that this would be called *iff* we already checked the shard's source and target node and that a replacement is ongoing, so that we don't need to check it there as well?

Reply:
(your other comments make it sound like that is the case)

Reply:
Yeah, that was my thought initially. I now wonder if the `NodeReplacementAllocationDecider` could be responsible for that check or at least the target part of the check. `BalancedShardsAllocator` could do the source check to only call this while a shard is really being replace-vacated.

Reply:
I've changed the `NodeReplacementAllocationDecider` to do this check I think in the way you meant.